app/Platform/Actions/UpdatePlayerGameMetricsAction.php

```diff
@@ -104,11 +154,156 @@ public function execute(PlayerGame $playerGame): void
             'first_unlock_hardcore_at' => $firstUnlockHardcoreAt,
             'points_total' => $pointsTotal,
             'points' => $points,
+            'points_hardcore' => $pointsHardcore,
             'points_weighted_total' => $pointsWeightedTotal,
             'points_weighted' => $pointsWeighted,
             'created_at' => $createdAt,
-        ];
+        ]);
+
+        PlayerGameMetricsUpdated::dispatch($user, $game);
+    }
+
+    private function checkCompletionBadge(User $user, Game $game, bool $hardcore): void
+    {
+        $data = getUnlockCounts($game->id, $user->username);
+
+        $minToCompleteGame = 6;
+        if ($data['NumAch'] >= $minToCompleteGame) {
+            $awardBadge = null;
+            if ($hardcore && $data['NumAwardedHC'] === $data['NumAch']) {
+                // all hardcore achievements unlocked, award mastery
+                $awardBadge = UnlockMode::Hardcore;
+            } elseif ($data['NumAwardedSC'] === $data['NumAch']) {
+                if ($hardcore && $this->playerBadgeExists($user->username, AwardType::Mastery, $game->id, UnlockMode::Softcore)) {
+                    // when unlocking a hardcore achievement, don't update the completion
+                    // date if the user already has a completion badge
+                } else {
+                    $awardBadge = UnlockMode::Softcore;
+                }
+            }
+
+            if ($awardBadge !== null) {
+                if (!$this->playerBadgeExists($user->username, AwardType::Mastery, $game->id, $awardBadge)) {
+                    $badge = AddSiteAward($user->username, AwardType::Mastery, $game->id, $awardBadge);
+
+                    PlayerBadgeAwarded::dispatch($badge);
+                    PlayerGameCompleted::dispatch($user, $game->id);
+
+                    if ($awardBadge === UnlockMode::Hardcore) {
+                        static_addnewhardcoremastery($game->id, $user->username);
+                    }
+                }
+
+                if (!RecentlyPostedProgressionActivity($user->username, $game->id, $awardBadge, ActivityType::CompleteGame)) {
+                    postActivity($user->username, ActivityType::CompleteGame, $game->id, $awardBadge);
+                }
+
+                expireGameTopAchievers($game->id);
+            }
+        } else {
+            $query = $user->playerBadges()
+                ->where('AwardType', AwardType::Mastery)
+                ->where('AwardData', $game->id);
+
+            if ($query->exists()) {
+                // TODO
+                $user->playerBadges()
+                    ->where('AwardType', AwardType::Mastery)
+                    ->where('AwardData', $game->id)
+                    ->delete();
```

Comment:
```suggestion
            $query = $user->playerBadges()
                ->where('AwardType', AwardType::Mastery)
                ->where('AwardData', $game->id)
                ->delete();
```

Pretty sure we can get rid of the double query and just call `delete()`, and if the record doesn't exist everything will be fine.

Reply:
That part isn't quite finished yet, I wasn't sure if the else branch actually follows the right condition yet after copying it over.